src/cast_core/src/CrudActor.cpp

```diff
@@ -1206,21 +1294,129 @@ struct CrudActor::PhaseConfig {
+                "Operation", "Operations", addOperation);
+        } else if (phaseContext["States"]) {  // Parse out the states}
+            if (!phaseContext["InitialStates"]) {
+                BOOST_THROW_EXCEPTION(InvalidConfigurationException(
+                    "CrudActor has States, but has not specified InitialStates"));
+            }
+            BOOST_LOG_TRIVIAL(debug) << "Adding states";
+            // build the list of states, and then actuall process the states
+            int i = 0;
+            std::unordered_map<std::string, int> stateNames;
+            for (auto [k, state] : phaseContext["States"]) {
+                BOOST_LOG_TRIVIAL(debug) << "Adding state to state Names " << state;
+                stateNames.emplace(state["Name"].to<std::string>(), i);
+                i++;
+            }
+            auto numStates = stateNames.size();
+            states.reserve(numStates);
+            for (auto [k, state] : phaseContext["States"]) {
+                BOOST_LOG_TRIVIAL(debug) << "Adding state " << state;
+                states.emplace_back(addState(state, stateNames));
+            }
+            // initial_state_weights.reserve(numStates);
+            initial_state_weights = std::vector<double>(numStates, 0);
+            // std::fill_n(initial_state_weights.begin(), numStates, 0);
+            for (auto [k, state] : phaseContext["InitialStates"]) {
+                initial_state_weights[stateNames[state["State"].to<std::string>()]] +=
+                    state["Weight"].to<double>();
+            }
+            BOOST_LOG_TRIVIAL(debug)
+                << "Length of initial state weights is " << initial_state_weights.size();
+            for (auto weight : initial_state_weights) {
+                BOOST_LOG_TRIVIAL(debug) << "Weight is " << weight;
+            }
+            BOOST_LOG_TRIVIAL(debug) << "Done adding states";
+        } else {  // Throw a useful error}
+            BOOST_THROW_EXCEPTION(
+                InvalidConfigurationException("CrudActor has neither Operations nor States "
+                                              "specified. Exactly one must be defined."));
+        }
     }
 };
 
 void CrudActor::run() {
+    int current_state = 0;
+    int next_state = 0;
     for (auto&& config : _loop) {
         auto session = _client->start_session();
+        // TODO: If running without states, define a single state with the operations
+        BOOST_LOG_TRIVIAL(debug) << "In CrudActor::run" << config->states.empty()
+                                 << config->continue_current_state;
+        if (!config->states.empty() and !config->continue_current_state) {
+            // pick the initial state for the phase
+            auto initial_distribution =
+                boost::random::discrete_distribution(config->initial_state_weights);
+            next_state = initial_distribution(_rng);
+            BOOST_LOG_TRIVIAL(debug) << "Picking initial state " << next_state;
+        }
```

Comment:
consider putting this logic into the `PhaseConfig` itself which will then potentially let you pull out the `StateManager` or whatever as a sub-member as mentioned in an earlier comment.